lib/aggregators/BaseAggregator.ts

```diff
@@ -43,9 +49,59 @@ export abstract class BaseAggregator<State> {
     return undefined;
   }
 
-  abstract init(start: RDF.Term): State;
+  protected abstract subInit(start: RDF.Term | undefined): SubState;
+  protected abstract subPut(state: SubState, bindings: RDF.Term | undefined): SubState;
+  protected abstract subResult(state: SubState): RDF.Term;
+
+  public init(start: RDF.Term | undefined, variable: string): IBaseState & { sub: SubState } {
+    const subState: SubState = this.subInit(start);
+    return {
+      sub: subState,
+      seen: this.addSeen({}, start, variable),
+    };
+  }
+
+  public put(state: IBaseState & { sub: SubState }, bindings: RDF.Term | undefined, variable: string):
+  IBaseState & { sub: SubState } {
+    if (this.canSkip(state, bindings, variable)) {
+      return state;
+    }
+    const sub: SubState = this.subPut(state.sub, bindings);
+    return { sub, seen: this.addSeen(state.seen, bindings, variable) };
+  }
+
+  public result(state: IBaseState & { sub: SubState }): RDF.Term {
+    return this.subResult(state.sub);
+  }
+
+  protected canSkip(state: IBaseState, term: RDF.Term | undefined, variable: string): boolean {
+    return this.distinct && (state.seen[variable] || []).some((t: RDF.Term) => t.equals(term));
+  }
+
+  protected addSeen(seen: Seen, term: RDF.Term | undefined,
+    variable: string): Seen {
+    return this.distinct ?
+      {
+        ...seen,
+        [variable]: [ ...seen[variable] || [], term ],
+      } :
+      {};
+  }
+}
+
+export abstract class SimpleAggregator<SubState> extends BaseAggregator<SubState> {
```

Comment:
I'm not a big fan of the `sub...` method names. Feels a bit hacky.

Instead of extension, I would suggest using object composition instead, which should solve this problem.

Concretely, the `BaseAggregator` can accept an `IAggregatorComponent` in its constructor, which has the old `BaseAggregator` interface.
Then, the `Count`, `Max`, ... classes can just be implementations of `IAggregatorComponent`.
This makes it possible to still use the old `init`, `put`, and `result` method names.

Instantiations of the different aggregators will then look something like this: `new BaseAggregator(new Count())`.
Since `BaseAggregator` won't be an abstract class anymore, we could even name it `Aggregator` now.